dali/python/nvidia/dali/pipeline.py

```diff
@@ -990,3 +996,63 @@ def iter_setup(self):
         For example, one can use this function to feed the input
         data from NumPy arrays."""
         pass
+
+
+def _discriminate_args(**kwargs):
+    """Split args on those applicable to Pipeline constructor and the rest."""
+    fca = inspect.getfullargspec(Pipeline.__init__)  # Full Ctor Args
+    ctor_args = dict(filter(lambda x: x[0] in fca.args or x[0] in fca.kwonlyargs, kwargs.items()))
+    fn_args = dict(filter(lambda x: x[0] not in ctor_args.keys(), kwargs.items()))
+    return ctor_args, fn_args
+
+
+def pipeline(fn=None, **pipeline_kwargs):
+    """
+    Decorator for wrapping functions that define DALI pipelines.
+
+    **Usage**
```

Comment:
I think that some description should go before usage/example:
```
    ``pipeline_def`` returns a function that creates a pipeline, where the processing graph is defined by the function ``fn`` passed as the first positional argument. It can be used as a decorator, in which case the decorated function becomes a pipeline factory.
The function returned is roughly equivalent to::

    def factory(*args, **kwargs)
        pipe = nvidia.dali.Pipeline()
        with pipe:
            pipe.set_outputs(*fn(*args, **kwargs))
        return pipe
```

Reply:
I believe that all this is written when you google "how to write a decorator?". So I wouldn't add it at the beginning, but I added it at the end of the paragraph, cause this note is in fact valuable.

Reply:
Sorry, but.... *WTF*? How can you possibly google that a decorator creates a DALI pipeline, calls your function within the with-scope of the newly created pipeline and calls set_outputs with the return values of the function you supplied???

Reply:
I think the doc is fine, we shouldn't use multi compound statements, it is hard to follow.

I was thinking whether to move the equivalent code to the top, but I'm not sure if it's better than usage examples.